WIP: Update visual plugin and add example

iche033 opened the pull request:
Made updates to get the visual plugin working. I also added an example `fresnel_plane` model and `fresnel.world`.

To test, just copy the `fresnel_plane` model to ~/.gazebo/models and run the example world:

```
GAZEBO_PLUGIN_PATH=${GAZEBO_PLUGIN_PATH}:`pwd`/build gazebo --verbose fresnel.world
```
The reflection is dark - that needs to be fixed.

![reflection](https://user-images.githubusercontent.com/4000684/63310537-ffe87300-c2af-11e9-9586-c479e45e1bc0.jpg)

iche033 committed:
update standalone example, add fresnel plane model

iche033 committed:
fixes

iche033 committed:
update refraction rt

tylerlum commented on ocean_visual.cc:
```diff
@@ -107,65 +105,121 @@ namespace asv
+    if (userCamera)
+    {
+      this->data->trackedCamera = userCamera;
+      this->data->camera =
```
I tried this a few days ago, but had issues. For some reason, when I set `camera->setPosition(user_camera->getPosition())` and `camera->setOrientation(user_camera->getOrientation())`, it was completely off. I tested this by having a miniscreen (Ogre::Rectangle2d) that displayed the texture created from the new camera, but it was looking completely off. I am not sure if that issue is relevant here.

Also, I got confused because this->data->scene->GetUserCamera(0)->WorldRotation() != this->data->scene->GetUserCamera(0)->OgreCamera()->getDerivedOrientation(). Is this expected?

tylerlum replied:
Is there a reason why you need to make the new camera? I tried it out, but it lead to more issues than I saw benefits.

iche033 replied:
`user_camera->getPosition()` and `user_camera->getOrientation()` gets the local position and orientation relative to the parent but we need to the derived position and orientation instead

It is expected that `WorldRotation != OgreCamera()->getDerivedOrientation`. It's a little confusing but `WorldRotation` returns the camera's parent node's world rotation, not the actual camera itself, i.e. `WorldRotation = OgreCamera()->getParentSceneNode()->getDerivedOrientation()`

 The camera object is rotated to be z-up when it's created: https://bitbucket.org/osrf/gazebo/src/2b59492f96fcf121b33f290fd5cf0b70b42f4e22/gazebo/rendering/Camera.cc#lines-1501

Having said all this, I think the new camera is probably not needed. We can remove it to simplify things a bit. I added it because I thought we needed to modify some properties and but looks like it's not needed.

iche033 replied:
Regarding the dark plane issue. I'm guessing it's because I forgot to set the viewport's background color. It defaults to black so it's probably reflecting the black background color. I need to verify this.

tylerlum replied:
I tried adding `vp->setBackgroundColour(Ogre::ColourValue::White);` but it is still Black. This is very strange.

tylerlum reviewed (commented):
Looks good to me. I am not sure why it ends up dark. I tried adding an Ogre light source, but it made no difference. I have a sense that the camera's render-to-texture is not working correctly, which is making it all black.

tylerlum commented on fresnel_plane/materials/scripts/fresnel.material:
```diff
@@ -0,0 +1,68 @@
+      texture_unit
+      {
+        // Will be filled in at runtime
+        // texture reflection
```
This is an interesting approach. In the Fresnel example, they have it not commented out. However for our use case, it may be that we need to define it later like this. I have applied this to the vrx PR

tylerlum commented on fresnel_plane/materials/scripts/fresnel_fp.glsl:
```diff
@@ -0,0 +1,41 @@
+#version 150
```
Are we able to use versoin 150 with Gazebo? Does it matter?

iche033 replied:
yeah version 150 works but probably not compatible with the waves ocean model

tylerlum commented on fresnel_plane/materials/scripts/fresnel_vp.glsl:
```diff
@@ -0,0 +1,37 @@
+in vec4 position;
+in vec3 normal;
+
+uniform mat4 worldViewProjMatrix;
```
In the current wave_gazebo, it has predefined variables like gl_ModelViewProjectionMatrix. This may just be because this is done in #version 150. Same with using `in` and `out` instead of `varying`.

iche033 replied:
yep I'm doing it all in version 150 for now to get something working first. We can backport to 120 if needed.

tylerlum commented:
I am headed home for the day. Thank you for your help and I hope we can get a working reflection model running soon!

iche033 committed:
fix dark plane

iche033 commented:
Fixed dark plane by setting background color. Also removed `this->data->trackedCamera`. 
4f15031

Results look better:

![reflection2](https://user-images.githubusercontent.com/4000684/63371596-2f8f8d80-c339-11e9-9ce4-8fe1539ce3bb.jpg)

The only TODO item left is to hide the objects above and below the plane depending on the render target being updated

tylerlum commented:
It looks great! I like how you set the background color to match, instead of just being black. I see there is some "refraction" of the cube right now, that is probably just what you mentioned about hiding objects above/below the plane depending on the render target

tylerlum commented:
I just tried out using a clip near plane for the reflection and refraction. I will add a commit to show it. I believe it fixed the problem.

tylerlum committed:
Use planeUp and planeDown as clipPlanes to hide objects for reflection/refraction rendering

tylerlum commented:
**NEW (in commit ee14b69)**

![image](https://user-images.githubusercontent.com/26510814/63373837-797a7280-c33d-11e9-961e-b270b414f29c.png)**
No refraction of above surface objects

![image](https://user-images.githubusercontent.com/26510814/63373892-931bba00-c33d-11e9-906d-823eec7d3fa3.png)
No reflection of below surface objects

**OLD**
![image](https://user-images.githubusercontent.com/26510814/63373980-c3fbef00-c33d-11e9-96f1-66261633b0e1.png)
Refraction of above surface objects

![image](https://user-images.githubusercontent.com/26510814/63374044-e1c95400-c33d-11e9-87b0-e2f63aea2628.png)
Reflection of below surface objects

iche033 commented:
nice! better than iterating through all the objects!

tylerlum commented:
@iche033 Is this good to be merged in?

iche033 commented:
yep looks good to me!
